### ISSUE TITLE
Adding/Changing http headers in Magento backend

### ISSUE THREAD

**Issue by ADDISON74** [nexcess/magento-turpentine#1014]
At the end of generated VCL file there are a lot unset statements. For misleading basic visitors, or those who are evaluating our website I am setting up some fake headers. The only way to do this is editing the generated VCL file. But when I am doing some changes in backend and saving the file, these changes are gone. 

Here is Turpentine default:

unset resp.http.X-Varnish;
unset resp.http.Via;
<strong>unset resp.http.X-Powered-By;</strong>
<strong>unset resp.http.Server;</strong>
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;

... and here is mine:

unset resp.http.X-Varnish;
unset resp.http.Via;
<strong>set resp.http.X-Powered-By = "HHVM"</strong>;
<strong>set resp.http.Server = "Lightspeed" </strong>;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;

Could you please allow editing this in backend and having them in place with every save of the file?

Thank you.

**Comment by aricwatson** [comment]
@ADDISON74 You should be able to accomplish this using the [Custom VCL File option](https://github.com/nexcess/magento-turpentine/wiki/Configuration#servers) - try creating a file in the location specified in the back-end with this code:

```
sub vcl_deliver { 
        set resp.http.X-Powered-By = "HHVM";
        set resp.http.Server = "Lightspeed" ;
}
```

Then, click 'Apply Varnish Config' under *Cache Management* in the back-end to load the new VCL. In my testing, the headers specified now stay visible.

**Comment by ADDISON74** [comment]
Thank you @aricwatson. I will follow as you recommended. It should work. Coming back to you with a feedback.

**Comment by ADDISON74** [comment]
I tried but it is not working (Turpentine dev latest). Here are the changes I've made:

Custom VCL File Location:
{{root_dir}}/app/code/community/Nexcessnet/Turpentine/misc/custom_include.vcl

custom_include.vcl content:
<code>sub vcl_deliver {
        set resp.http.X-Powered-By = "HHVM";
        set resp.http.Server = "Lightspeed" ;
}</code>

I saved and applied Varnish Config in Cache Management backend section, I flushed Magento cache too. Visiting the frontend it is not getting these two headers.

I checked Varnish/Turpentine configuration file (default.vcl). Well custom_include.vcl content it is placed into this file between "import std" and  "backend default" statements.

I went further, in default.vcl at the end I commented these two lines added by Turpentine in else statement:

<code> else { ... #unset resp.http.X-Powered-By;
#unset resp.http.Server; ... }}
</code>

This means changing file manually and a save action in backend will overwrite these changes. As I expected now headers I set up inside the custom file are appearing in place.

In conclusion using a custom file isn't providing the correct results. This needs an investigation from you. Please let me know when you find something.

**Comment by aricwatson** [comment]
Hmm, just re-tested and I am seeing the headers specified in custom-include.vcl without any need to change the default.vcl.

This is using the current devel branch of Turpentine, with Varnish 3.0.4 - what version of Varnish are you using?

**Comment by ADDISON74** [comment]
I am using Varnish-4.0.3 with the latest devel version.

1) Magento backend System -> Configuration -> Varnish Options:
Custom VCL File Location 
{{root_dir}}/app/code/community/Nexcessnet/Turpentine/misc/custom_include.vcl

2)  app/code/community/Nexcessnet/Turpentine/misc/custom_include.vcl. This files has the following content: 

sub vcl_deliver {
    set resp.http.X-Powered-By = "HHVM";
    set resp.http.Server = "Lightspeed" ;
}

3) In Cache Management I [Save Config] then [Apply Config] and [Download Config] (download content I will show you in the next post.

Let's reload the frontend in browser and check the page headers:

Remote Address:192.168.159.100:80
Request URL:http://www.mydomain.com/
Request Method:GET
Status Code:200 OK

Age:8
Connection:keep-alive
Content-Encoding:gzip
Content-Type:text/html; charset=UTF-8
Date:Mon, 23 Nov 2015 21:21:54 GMT
Transfer-Encoding:chunked
X-Frame-Options:SAMEORIGIN

Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-US,en;q=0.8
Connection:keep-alive
Cookie:frontend=5g3a4b77p4ptbuddgn60ks6ip1; PHPSESSID=4a640pqmk2a0k8hplj94uagtd1
Host:www.mmsport.ro
Referer:http://www.mydomain.com/
Upgrade-Insecure-Requests:1
User-Agent:Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36

There are no Server or X-Powered-By headers!

**Comment by ADDISON74** [comment]
Here is the default.vcl file content I downloaded from Cache Management. Please pay attention where is custom_include.vcl file content inserted (bold paragraph):

vcl 4.0;
C{
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <pthread.h>
static pthread_mutex_t lrand_mutex = PTHREAD_MUTEX_INITIALIZER;
void generate_uuid(char* buf) {
pthread_mutex_lock(&lrand_mutex);
long a = lrand48();
long b = lrand48();
long c = lrand48();
long d = lrand48();
pthread_mutex_unlock(&lrand_mutex);
sprintf(buf, "frontend=%08lx%04lx%04lx%04lx%04lx%08lx",
a,
b & 0xffff,
(b & ((long)0x0fff0000) >> 16) | 0x4000,
(c & 0x0fff) | 0x8000,
(c & (long)0xffff0000) >> 16,
d
);
return;
}
}C
import std;
<strong>sub vcl_deliver {
set resp.http.X-Powered-By = "HHVM";
set resp.http.Server = "Lightspeed" ;
}</strong>
backend default {
.host = "192.168.159.100";
.port = "8080";
.first_byte_timeout = 300s;
.between_bytes_timeout = 300s;
}
backend admin {
.host = "192.168.159.100";
.port = "8080";
.first_byte_timeout = 21600s;
.between_bytes_timeout = 21600s;
}
acl crawler_acl {
"127.0.0.1";
}
acl debug_acl {
}
/* -- REMOVED
sub generate_session {
if (req.url ~ ".*[&?]SID=([^&]+).*") {
set req.http.X-Varnish-Faked-Session = regsub(
req.url, ".*[&?]SID=([^&]+).*", "frontend=\1");
} else {
C{
char uuid_buf [50];
generate_uuid(uuid_buf);
static const struct gethdr_s VGC_HDR_REQ_VARNISH_FAKED_SESSION =
{ HDR_REQ, "\030X-Varnish-Faked-Session:"};
VRT_SetHdr(ctx,
&VGC_HDR_REQ_VARNISH_FAKED_SESSION,
uuid_buf,
vrt_magic_string_end
);
}C
}
if (req.http.Cookie) {
std.collect(req.http.Cookie);
set req.http.Cookie = req.http.X-Varnish-Faked-Session +
"; " + req.http.Cookie;
} else {
set req.http.Cookie = req.http.X-Varnish-Faked-Session;
}
}
sub generate_session_expires {
C{
time_t now = time(NULL);
struct tm now_tm = *gmtime(&now);
now_tm.tm_sec += 3600;
mktime(&now_tm);
char date_buf [50];
strftime(date_buf, sizeof(date_buf)-1, "%a, %d-%b-%Y %H:%M:%S %Z", &now_tm);
static const struct gethdr_s VGC_HDR_RESP_COOKIE_EXPIRES =
{ HDR_RESP, "\031X-Varnish-Cookie-Expires:"};
VRT_SetHdr(ctx,
&VGC_HDR_RESP_COOKIE_EXPIRES,
date_buf,
vrt_magic_string_end
);
}C
}
-- */
sub vcl_recv {
if (req.restarts == 0) {
if (req.http.X-Forwarded-For) {
set req.http.X-Forwarded-For =
req.http.X-Forwarded-For + ", " + client.ip;
} else {
set req.http.X-Forwarded-For = client.ip;
}
}
if (!true || req.http.Authorization ||
req.method !~ "^(GET|HEAD|OPTIONS)$" ||
req.http.Cookie ~ "varnish_bypass=1") {
return (pipe);
}
if(false) {
set req.http.X-Varnish-Origin-Url = req.url;
}
set req.url = regsuball(req.url, "(.*)//+(.*)", "\1/\2");
if (req.http.Accept-Encoding) {
if (req.http.Accept-Encoding ~ "gzip") {
set req.http.Accept-Encoding = "gzip";
} else if (req.http.Accept-Encoding ~ "deflate") {
set req.http.Accept-Encoding = "deflate";
} else {
unset req.http.Accept-Encoding;
}
}
if (req.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?") {
set req.http.X-Turpentine-Secret-Handshake = "1";
if (req.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?admin") {
set req.backend_hint = admin;
return (pipe);
}
if (req.http.Cookie ~ "\bcurrency=") {
set req.http.X-Varnish-Currency = regsub(
req.http.Cookie, ".*\bcurrency=([^;]*).*", "\1");
}
if (req.http.Cookie ~ "\bstore=") {
set req.http.X-Varnish-Store = regsub(
req.http.Cookie, ".*\bstore=([^;]*).*", "\1");
}
if (req.url ~ "/turpentine/esi/get(?:Block|FormKey)/") {
set req.http.X-Varnish-Esi-Method = regsub(
req.url, ".*/method/(\w+)/.*", "\1");
set req.http.X-Varnish-Esi-Access = regsub(
req.url, ".*/access/(\w+)/.*", "\1");
if (req.http.X-Varnish-Esi-Method == "esi" && req.esi_level == 0 &&
!(false || client.ip ~ debug_acl)) {
return (synth(403, "External ESI requests are not allowed"));
}
}
if (req.http.Cookie !~ "frontend=" && !req.http.X-Varnish-Esi-Method) {
if (client.ip ~ crawler_acl ||
req.http.User-Agent ~ "^(?:ApacheBench/.*|.*Googlebot.*|JoeDog/.*Siege.*|magespeedtest\.com|Nexcessnet_Turpentine/.*)$") {
set req.http.Cookie = "frontend=crawler-session";
} else {
return (pipe);
}
}
if (true &&
req.url ~ ".*\.(?:css|js|jpe?g|png|gif|ico|swf)(?=\?|&|$)") {
unset req.http.Cookie;
unset req.http.X-Varnish-Faked-Session;
return (hash);
}
if (req.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?(?:admin|api|cron\.php)" ||
req.url ~ "\?.*__from_store=") {
return (pipe);
}
if (true &&
req.url ~ "(?:[?&](?:__SID|XDEBUG_PROFILE)(?=[&=]|$))") {
return (pass);
}
if (req.url ~ "[?&](utm_source|utm_medium|utm_campaign|gclid|cx|ie|cof|siteurl)=") {
set req.url = regsuball(req.url, "(?:(\?)?|&)(?:utm_source|utm_medium|utm_campaign|gclid|cx|ie|cof|siteurl)=[^&]+", "\1");
set req.url = regsuball(req.url, "(?:(\?)&|\?$)", "\1");
}
if (true && req.url ~ "[?&](utm_source|utm_medium|utm_campaign|utm_content|utm_term|gclid|cx|ie|cof|siteurl)=") {
set req.url = regsuball(req.url, "(?:(\?)?|&)(?:utm_source|utm_medium|utm_campaign|utm_content|utm_term|gclid|cx|ie|cof|siteurl)=[^&]+", "\1");
set req.url = regsuball(req.url, "(?:(\?)&|\?$)", "\1");
}
if(false) {
set req.http.X-Varnish-Cache-Url = req.url;
set req.url = req.http.X-Varnish-Origin-Url;
unset req.http.X-Varnish-Origin-Url;
}
return (hash);
}
}
sub vcl_pipe {
unset bereq.http.X-Turpentine-Secret-Handshake;
set bereq.http.Connection = "close";
}
sub vcl_hash {
if(false && req.http.X-Varnish-Cache-Url) {
hash_data(req.http.X-Varnish-Cache-Url);
} else {
hash_data(req.url);
}
if (req.http.Host) {
hash_data(req.http.Host);
} else {
hash_data(server.ip);
}
hash_data(req.http.Ssl-Offloaded);
if (req.http.X-Normalized-User-Agent) {
hash_data(req.http.X-Normalized-User-Agent);
}
if (req.http.Accept-Encoding) {
hash_data(req.http.Accept-Encoding);
}
if (req.http.X-Varnish-Store || req.http.X-Varnish-Currency) {
hash_data("s=" + req.http.X-Varnish-Store + "&c=" + req.http.X-Varnish-Currency);
}
if (req.http.X-Varnish-Esi-Access == "private" &&
req.http.Cookie ~ "frontend=") {
hash_data(regsub(req.http.Cookie, "^.*?frontend=([^;]*);*.*$", "\1"));
}
return (lookup);
}
sub vcl_hit {
}
sub vcl_backend_response {
set beresp.grace = 15s;
set beresp.http.X-Varnish-Host = bereq.http.host;
set beresp.http.X-Varnish-URL = bereq.url;
if (bereq.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?") {
unset beresp.http.Vary;
set beresp.do_gzip = true;
if (beresp.status != 200 && beresp.status != 404) {
set beresp.ttl = 15s;
set beresp.uncacheable = true;
return (deliver);
} else {
if (beresp.http.Set-Cookie) {
set beresp.http.X-Varnish-Set-Cookie = beresp.http.Set-Cookie;
unset beresp.http.Set-Cookie;
}
unset beresp.http.Cache-Control;
unset beresp.http.Expires;
unset beresp.http.Pragma;
unset beresp.http.Cache;
unset beresp.http.Age;
if (beresp.http.X-Turpentine-Esi == "1") {
set beresp.do_esi = true;
}
if (beresp.http.X-Turpentine-Cache == "0") {
set beresp.ttl = 15s;
set beresp.uncacheable = true;
return (deliver);
} else {
if (true &&
bereq.url ~ ".*\.(?:css|js|jpe?g|png|gif|ico|swf)(?=\?|&|$)") {
set beresp.ttl = 604800s;
set beresp.http.Cache-Control = "max-age=604800";
} elseif (bereq.http.X-Varnish-Esi-Method) {
if (bereq.http.X-Varnish-Esi-Access == "private" &&
bereq.http.Cookie ~ "frontend=") {
set beresp.http.X-Varnish-Session = regsub(bereq.http.Cookie,
"^.*?frontend=([^;]*);*.*$", "\1");
}
if (bereq.http.X-Varnish-Esi-Method == "ajax" &&
bereq.http.X-Varnish-Esi-Access == "public") {
set beresp.http.Cache-Control = "max-age=" + regsub(
bereq.url, ".*/ttl/(\d+)/.*", "\1");
}
set beresp.ttl = std.duration(
regsub(
bereq.url, ".*/ttl/(\d+)/.*", "\1s"),
300s);
if (beresp.ttl == 0s) {
set beresp.ttl = 15s;
set beresp.uncacheable = true;
return (deliver);
}
} else {
set beresp.ttl = 3600s;
}
}
}
return (deliver);
}
}
sub vcl_deliver {
if (req.http.X-Varnish-Faked-Session) {
set resp.http.Set-Cookie = req.http.X-Varnish-Faked-Session +
"; expires=" + resp.http.X-Varnish-Cookie-Expires + "; path=/";
if (req.http.Host) {
if (req.http.User-Agent ~ "^(?:ApacheBench/.*|.*Googlebot.*|JoeDog/.*Siege.*|magespeedtest\.com|Nexcessnet_Turpentine/.*)$") {
set resp.http.Set-Cookie = resp.http.Set-Cookie +
"; domain=" + regsub(req.http.Host, ":\d+$", "");
} else {
if(req.http.Host ~ "") {
set resp.http.Set-Cookie = resp.http.Set-Cookie +
"; domain=";
} else {
set resp.http.Set-Cookie = resp.http.Set-Cookie +
"; domain=" + regsub(req.http.Host, ":\d+$", "");
}
}
}
set resp.http.Set-Cookie = resp.http.Set-Cookie + "; httponly";
unset resp.http.X-Varnish-Cookie-Expires;
}
if (req.http.X-Varnish-Esi-Method == "ajax" && req.http.X-Varnish-Esi-Access == "private") {
set resp.http.Cache-Control = "no-cache";
}
if (false || client.ip ~ debug_acl) {
set resp.http.X-Varnish-Hits = obj.hits;
set resp.http.X-Varnish-Esi-Method = req.http.X-Varnish-Esi-Method;
set resp.http.X-Varnish-Esi-Access = req.http.X-Varnish-Esi-Access;
set resp.http.X-Varnish-Currency = req.http.X-Varnish-Currency;
set resp.http.X-Varnish-Store = req.http.X-Varnish-Store;
} else {
unset resp.http.X-Varnish;
unset resp.http.Via;
unset resp.http.X-Powered-By;
unset resp.http.Server;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;
}
}

As you can custom file content is inserted at the top of default.vcl file content and these headers are canceled by the ones at the end of the file. Turpentine is placing custom file content in a wrong place.

**Comment by ADDISON74** [comment]
Now let move custom file content from its actual position inside default.vcl file at the end of the file. Here is the last part of default.vcl:

else {
unset resp.http.X-Varnish;
unset resp.http.Via;
unset resp.http.X-Powered-By;
unset resp.http.Server;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;
}
}
<strong>sub vcl_deliver {
set resp.http.X-Powered-By = "HHVM";
set resp.http.Server = "Lightspeed" ;
}</strong>

Let's check again the frontend in browser. Here are the Headers:

Remote Address:192.168.159.100:80
Request URL:http://www.mydomain.com/
Request Method:GET
Status Code:200 OK

Age:0
Connection:keep-alive
Content-Encoding:gzip
Content-Type:text/html; charset=UTF-8
Date:Mon, 23 Nov 2015 21:32:19 GMT
<strong>Server:Lightspeed</strong>
Transfer-Encoding:chunked
X-Frame-Options:SAMEORIGIN
<strong>X-Powered-By:HHVM</strong>

Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-US,en;q=0.8
Cache-Control:max-age=0
Connection:keep-alive
Cookie:frontend=5g3a4b77p4ptbuddgn60ks6ip1; PHPSESSID=4a640pqmk2a0k8hplj94uagtd1
Host:www.mydomain.com
Referer:http://www.mydomain.com/
Upgrade-Insecure-Requests:1
User-Agent:Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36

As you can see those two headers set up in custom file are now in Headers. But we achieved this moving at the end the custom file content. There is one conclusion, you should place custom file content at the end of the default.vcl file.

**Comment by ADDISON74** [comment]
I am using Varnish 4. Inside this file <strong>/app/code/community/Nexcessnet/Turpentine/misc/varnish-4.vcl</strong> this statement is in a wrong place <strong>{{custom_vcl_include}}</strong>.

All custom codes must be set up at the end of this file in order to have priority. In my examples above, moving custom_vcl_include file content at the end of Varnish file headers appear as expected. If {{custom_vcl_include}} is still kept in the same place like it is right now, all the custom headers are canceled by the ones at the end of the file and this create my issue.

**Comment by aricwatson** [comment]
Ah! I found the problem - the reason I was *not* encountering the same problem (even when I tested just now with Varnish 4) is that I had debug turned *on* in Varnish - which means these weren **not** being unset:

```
unset resp.http.X-Powered-By;
unset resp.http.Server;
```

Sorry for being dense! And I agree, I think the custom_vcl_include should be placed at the bottom of the file, where it can potentially be used to override auto-generated VCL code from Turpentine. We'll test this a bit first and set it up if there are problems.

**Comment by ADDISON74** [comment]
Don't worry for being dense, I am glad finally you understood the issue. If all is OK in your tests do not forget to update Turpentine extension at once for all configuration files in /misc folder. In my only test related to those headers it worked as expected. But I don't know if other Varnish statements must be in other places. Please do an evaluation for different scenarios, but I guess custom code being at the end will be applied always for sure.

**Comment by aricwatson** [comment]
OK I've (finally) updated (*devel* branch) so that the custom included VCL is added at the *bottom* of the  generated VCL file. We'd appreciate extra testing, but so far so good in my testing.

**Comment by aricwatson** [comment]
Closing this issue, as this is (hopefully) fixed.
